### ISSUE TITLE
Server error on admin preview of VirtualSourceObject

### ISSUE THREAD

**Issue by relikd** [lektor/lektor#1016]
https://github.com/lektor/lektor/blob/9ede53d740e76f785babf81cb0cab5c92c9a16c5/lektor/admin/modules/api.py#L134-L141

Should a virtual object be treated differently by the admin UI in regards to the navigation redirect? The highlighted snippet above results in a `TypeError: object is not subscriptable` on the attributes `_path` and `_alt`. Further, if you provide a path, the admin UI will redirect to that path. Which is funny because the error happens while the correct page is being displayed. I believe the correct behaviour is to prevent any further navigation if `isinstance(record, VirtualSourceObject)`?

**Comment by dairiki** [comment]
I see two possible correct solutions (though there could be others):
1. As you suggest, return <del>`None`</del> <ins>`jsonify(exists=False, path=None, alt=None)`</ins> to keep the admin UI on whatever record it was on previously [edited: return `jsonify(exists=False, ...)`, not return `None`]
2. Return the path (and alt) of the parent _record_ of the _virtual source_ (available in the `VirtualSourceObject.record` attribute), so the admin UI will navigate to the parent record.

My gut feeling is that \#2 is the better option, but I guess it would depend somewhat on the details of specific virtual source object in question.

@relikd: Do you have a specific use-case where you're hitting this issue? Which of the two behaviors makes the most sense in that case?

**Comment by relikd** [comment]
I cannot return `None` because then my Virtual Object is not loaded, instead it will show a 401 not found. And No.2 is also not right because I do not want to navigate to the parent record, I want to display the content of my Virtual Object. Here is a short outline what happens:

1. open `/blog/` in edit mode
2. the template for `/blog` may have something like `{{ '/blog@vpath/further/link' | url }}`
3. my virtualpathresolver fixes the url to point to an actual html file, e.g. `/blog/artifical/path/`
4. my urlresolver makes a modification to fix `/index.html` at the end
5. open `/blog` in preview mode
6. click on the link which is pointing to `/blog/artifical/path/`
7. admin UI loads the correct Virtual Object but then the error appears

Now, if I were to return `None` instead, my Virtual Object would not load.
And if I were to return the path of the parent, admin UI would immediatelly redirect to `/blog` instead and I would not be able to preview the html of my Virtual Object.

**Comment by dairiki** [comment]
> I cannot return `None` because then my Virtual Object is not loaded, instead it will show a 401 not found.

I'm fairly sure that that is not true.

The use of the `/matchurl` API method is as follows:
When the Admin UI is in "preview" mode, the preview is displayed in an iframe embedded within the Admin UI.  When that iframe is navigated by clicking on a link within the iframe, the iframe navigates itself to the new page — no javascript involved.
The client-side javascript code notices when the iframe navigates to a new URL and then uses the `/matchurl` API to figure out how to update the rest of the Admin UI (e.g. what record gets edited with the "edit" button is clicked and which children and attachments are listed in the left side-bar.)

(If the call to `/matchurl` fails or returns garbage, the preview iframe will still work fine, but the rest of the Admin UI will either not update or get screwed up.)

I've just tested this code:
```py
@bp.route("/matchurl")
def match_url():
    record = g.admin_context.pad.resolve_url_path(
        request.args["url_path"], alt_fallback=False
    )
    # Resolve virtual source to parent record
    record = getattr(record, 'record')
    if record is None:
        return jsonify(exists=False, path=None, alt=None)
    return jsonify(exists=True, path=record["_path"], alt=record["_alt"])
```
and (in quick tests) it appears to work correctly.

**Comment by relikd** [comment]
My problem is that `/artificial/path/` is not automatically resolved, even though the file `/artificial/path/index.html` exists. This is why I cant return `None` – I need a custom resolver to "redirect" to the `index.html`. As I said, if I do return `None`, lektor wont resolve to the `/index.html`. I think the initial problem, that it is not resolved in the first place, comes from generating the virtual source. The call to declare artifact requires a `/index.html`:

```python
self.declare_artifact("/artificial/path/index.html", sources=[...])
```

I assume this is the reason why the server does not know how to resolve from the parent dir.

**Comment by relikd** [comment]
Here is a minimal example:

```python
# -*- coding: utf-8 -*-
from lektor.pluginsystem import Plugin
from lektor.build_programs import BuildProgram
from lektor.sourceobj import VirtualSourceObject
from lektor.utils import build_url


class TestSource(VirtualSourceObject):
    @property
    def path(self) -> str:
        return build_url([self.record.path, "@test", "hi"])

    @property
    def url_path(self) -> str:
        return build_url([self.record.path, "artificial", "hi"])


class TestProgram(BuildProgram):
    def produce_artifacts(self) -> None:
        self.declare_artifact(
            self.source.url_path + 'index.html',
            sources=list(self.source.record.iter_source_filenames())
        )

    def build_artifact(self, artifact) -> None:
        self.source.pad.db.track_record_dependency(self.source)
        artifact.render_template_into("view_source.html", this=self.source)


class Main(Plugin):
    def on_setup_env(self, **extra):
        self.env.add_build_program(TestSource, TestProgram)

        @self.env.urlresolver
        def _fn(node, pieces):
            if node.path == '/' and pieces == ['artificial', 'hi']:
                return TestSource(node)

    def on_before_build_all(self, builder, **extra):
        root = builder.pad.get_all_roots()[0]
        builder.build(TestSource(root))
```

Without the `urlresolver` you can open `/artificial/hi/index.html` but not `/artificial/hi/`. With the resolver you get the server error. (just add a link to your markdown `[test](/artificial/hi/)`)

**Comment by dairiki** [comment]
First, where I wrote "return `None`" in my [first comment](#issuecomment-1082357007) above, what I really meant (ahem... sorry) was "return `jsonify(exists=False, path=None, alt=None)`", as is currently done when `Pad.resolve_url_path` returns `None`.  (I've edited my comment to fix that. I apologize for the confusion.)

I haven't tried your test code, but, a couple of observations:

 Yes, you do need to register the `urlresolver` if you want the admin server to be able to correctly serve the virtual source's artifact at `http://localhost:5000/artificial/hi/`.  The urlresolver is the only way to resolve the requested URL to the (virtual) source object — though if URL resolution fails, the admin server falls back to looking in the output file tree (which is why `http://localhost:5000/artificial/hi/index.html` works).

I'm still not convinced that the `/matchurl` admin API method has anything to do with resolving the HTTP request.  As I noted above, it is used by the admin UI client-side javascript app to figure out where to point the rest of the admin interface to when the preview iframe is navigated.  Yes `/matchurl` is throwing an exception when it's `url_path` argument resolves to a virtual source object, and that is a bug, but that's not why the admin http server can or can not find `http://localhost:5000/artificial/hi/`.

The `TestSource.path` property should probably not be using `build_url`.  It should be returning a Lektor database path (suitable for passing to `Pad.get`): it should not have a trailing slash, and it should have no slash between the parent record path and the "@".  That said, I don't think that this has anything to with the URL resolution issues you're experiencing.  In any case, try something like:
```py
class TestSource(VirtualSourceObject):
    @property
    def path(self) -> str:
        return f"{self.record.path}@test/hi"
    ...
```

Note that if one wants to be able to resolve that path back to the virtual source object, then one must also declare a `virtualpathresolver`.  (Roughly, the `virtualpathresolver` is the inverse to `VirtualSourceObject.path`, the same way that the `urlresolver` is the inverse to `VirtualSourceObject.url_path`.)

**Comment by relikd** [comment]
Thanks for the hint on `build_url`. And back to the issue: no, the admin API does not resolve the HTTP request, but it tries to redirect it. My `urlresolver` and `virtualpathresolver` work fine, just the admin UI tries to redirect away from the already found virtual object.

I just tried to replace:
```python
if record is None:
    return jsonify(exists=False, path=None, alt=None) 
```
with:
```python
if record is None or isinstance(record, VirtualSourceObject):
    return jsonify(exists=False, path=None, alt=None) 
```

and the unwanted redirect / server error disappears. I would assume you never want to redirect if the record is a virtual object. But I do not know whether this may affect other parts I haven't thought of.

**Comment by dairiki** [comment]
> And back to the issue: no, the admin API does not resolve the HTTP request, but it tries to redirect it.

My apologies, you're right.

I think the proper fix is:
```py
if not isinstance(record, lektor.db.Record):
    return jsonify(exists=False, path=None, alt=None) 
```
because as it stands now, `/matchurl` also fails if given a URL to an `Asset`.  And (oh look) it turns out that this is fixed in 82d3e66f9ed9b6f93a372a86591b32a84b2e6e84 which is part of the still unmerged PR #987.

I will work on merging #987, which should fix this issue.

**Comment by relikd** [comment]
Thanks. Just one more question, are you sure its Record? Because a Page is also a Record. And doesn't the UI redirect to these?

**Comment by dairiki** [comment]
> Just one more question, are you sure its Record? Because a Page is also a Record. And doesn't the UI redirect to these?

I'm not sure I understand.  `Page` is a subclass of `Record` (as is `Attachment`, `Image`, etc.)  If `record` is an instance of `Page`, then `isinstance(record, Record)` will be true.  (A instance of a subclass is also an instance of any superclasses.)

**Comment by dairiki** [comment]
I've just merged #987, so this issue is fixed (I think) in the master branch.

The next release of the master branch will be 3.4.0 — and I'm not quite sure when that will be.  There are a couple more changes (API changes, dependency changes) I think we want to get in before we do that.

Is it worth back-porting this fix to the 3.3 branch (and making a 3.3.x release)?  Or can you work from the master branch for now?

**Comment by relikd** [comment]
> I'm not sure I understand. `Page` is a subclass of `Record` (as is `Attachment`, `Image`, etc.) If `record` is an instance of `Page`, then `isinstance(record, Record)` will be true. (A instance of a subclass is also an instance of any superclasses.)

Yes, that is why I ask. Should a Page not return `jsonify(exists=True,...` instead of `jsonify(exists=False,...`? I thought a Page should evaluate as it did before? Only virtual objects and assets cause problems, so these two should return `False` whereas a page should return `True`? -- Or: which records remain if you test for isinstance Record? When does it evaluate to true?

> Is it worth back-porting this fix to the 3.3 branch (and making a 3.3.x release)? Or can you work from the master branch for now?

No need for me, I'll update once its ready.

**Comment by dairiki** [comment]
> Should a Page not return `jsonify(exists=True,...` instead of `jsonify(exists=False,...`?

Indeed.  I think you maybe missed the `not` in front of `isinstance`. ;-)

> No need for me, I'll update once its ready.

Great!  I'll close this issue.
Feel free to re-open if there's trouble.
